flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java

```diff
@@ -640,50 +640,75 @@ private void addEdgeInternal(
-                }
-            }
+    private void createActualEdge(
+            Integer upStreamVertexID,
+            Integer downStreamVertexID,
+            int typeNumber,
+            StreamPartitioner<?> partitioner,
+            OutputTag outputTag,
+            StreamExchangeMode exchangeMode) {
+        StreamNode upstreamNode = getStreamNode(upStreamVertexID);
+        StreamNode downstreamNode = getStreamNode(downStreamVertexID);
+
+        // If no partitioner was specified and the parallelism of upstream and downstream
+        // operator matches use forward partitioning, use rebalance otherwise.
+        if (partitioner == null
+                && upstreamNode.getParallelism() == downstreamNode.getParallelism()) {
+            partitioner = new ForwardPartitioner<Object>();
+        } else if (partitioner == null) {
+            partitioner = new RebalancePartitioner<Object>();
+        }
 
-            if (exchangeMode == null) {
-                exchangeMode = StreamExchangeMode.UNDEFINED;
+        if (partitioner instanceof ForwardPartitioner) {
+            if (upstreamNode.getParallelism() != downstreamNode.getParallelism()) {
+                throw new UnsupportedOperationException(
+                        "Forward partitioning does not allow "
+                                + "change of parallelism. Upstream operation: "
+                                + upstreamNode
+                                + " parallelism: "
+                                + upstreamNode.getParallelism()
+                                + ", downstream operation: "
+                                + downstreamNode
+                                + " parallelism: "
+                                + downstreamNode.getParallelism()
+                                + " You must use another partitioning strategy, such as broadcast, rebalance, shuffle or global.");
             }
+        }
 
-            StreamEdge edge =
-                    new StreamEdge(
-                            upstreamNode,
-                            downstreamNode,
-                            typeNumber,
-                            partitioner,
-                            outputTag,
-                            exchangeMode);
-
-            getStreamNode(edge.getSourceId()).addOutEdge(edge);
-            getStreamNode(edge.getTargetId()).addInEdge(edge);
+        if (exchangeMode == null) {
+            exchangeMode = StreamExchangeMode.UNDEFINED;
         }
+
+        /**
+         * Just make sure that {@link StreamEdge} connecting same nodes (for example as a result of
+         * self unioning a {@link DataStream}) are distinct and unique. Otherwise it would be
+         * difficult on the {@link StreamTask} to assign {@link RecordWriter}s to correct {@link
+         * StreamEdge}.
+         */
+        int uniqueId = getStreamEdges(upstreamNode.getId(), downstreamNode.getId()).size();
```

Comment:
Since the `uniqueId` here is an integer, I guess it can be `int` instead of `long` also in the `StreamEdge`.
Moreover, is the `size()` of this list enough to uniquely identify the `StreamEdge`?
Wouldn't it be better to have an atomically incremented int instead?

Reply:
> is the size() of this list enough to uniquely identify the StreamEdge?
> Wouldn't it be better to have an atomically incremented int instead?

I've explained what type of uniqueness we need in the java doc of the field. `size()` is not enough to uniquely identify `StreamEdge`, but combined with `sourceId` and `targetId` it is. Regarding why no global incremented integer, I thought about it, but I wanted to have more self explanatory code, why do we need this uniqueId. That the tuple `sourceId`, `targetId`, + partitioning scheme, is not enough to uniquely distinguish edges. By using
```
getStreamEdges(upstreamNode.getId(), downstreamNode.getId()).size();

```
I'm ensuring almost exactly this (% partitioning schemes). That if there is more then one edge connecting `sourceId` and `targetId`, they will have different `uniqueId`.

Reply:
Thank you @pnowojski ! I understood the concept (and it's well explained in the javadoc), was a bit confused with the `size()` here, but yeah, on every call to size() there is already one more `StreamEdge` in the list so the `size()` increases, got it!